File: HedgeLib/IO/FileBase.cs
using System;
using System.IO;

namespace HedgeLib.IO
{
    /// <summary>
    /// The base for all file-based classes in HedgeLib.
    /// </summary>
    public class FileBase
    {
        // Methods
        public virtual void Load(string filePath)
        {
            // Throw exceptions if necessary
            if (string.IsNullOrEmpty(filePath))
                throw new ArgumentNullException("filePath");

            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException(
                    "The given file could not be loaded, as it cannot be found!", filePath);
            }

            // Load the file
            using (var fileStream = File.OpenRead(filePath))
            {
                Load(fileStream);
            }
        }

        public virtual void Load(Stream fileStream)
        {
            throw new NotImplementedException();
        }

        public virtual void Save(string filePath, bool overwrite = false)
        {
            // Throw exceptions if necessary
            if (string.IsNullOrEmpty(filePath))
                throw new ArgumentNullException("filePath");

            if (!overwrite && File.Exists(filePath))
                throw new Exception("Cannot save the given file - it already exists!");

            // Save the file
            using (var fileStream = File.Create(filePath))
            {
                Save(fileStream);
            }
        }

        public virtual void Save(Stream fileStream)
        {
            throw new NotImplementedException();
        }
    }
}

File: HedgeLib/src/IO/fileBase.cpp
#include "IO/fileBase.h"
#include "IO/file.h"
#include <filesystem>
#include <stdexcept>

namespace HedgeLib::IO
{
	FileBase::~FileBase() {}

	void FileBase::Load(const std::filesystem::path filePath)
	{
		Read(File::OpenRead(filePath));
	}

	void FileBase::Read(const File& file)
	{
		throw std::logic_error(
			"Cannot call Read function; function not yet implemented!");
	}

	void FileBase::Save(const std::filesystem::path filePath)
	{
		Write(File::OpenWrite(filePath));
	}

	void FileBase::Write(const File& file)
	{
		throw std::logic_error(
			"Cannot call Write function; function not yet implemented!");
	}
}